web/server/codechecker_server/api/report_server.py

```diff
@@ -2829,54 +2830,89 @@ def massStoreRun(self, name, tag, version, b64zip, force,
+                            # Load the lock record for "FOR UPDATE" so that the
+                            # transaction that handles the run's store
+                            # operations has a lock on the database row itself.
+                            run_lock = session.query(RunLock) \
+                                .filter(RunLock.name == name) \
+                                .with_for_update(nowait=True).one()
+
+                            # Do not remove this seemingly dummy print, we need
+                            # to make sure that the execution of the SQL
+                            # statement is not optimised away and the fetched
+                            # row is not garbage collected.
+                            LOG.debug("Storing into run '%s' locked at '%s'.",
+                                      name, run_lock.locked_at)
+
+                            # Actual store operation begins here.
+                            user_name = self.__get_username()
+                            run_id = \
+                                store_handler.addCheckerRun(session,
+                                                            command,
+                                                            name,
+                                                            tag,
+                                                            user_name,
+                                                            run_history_time,
+                                                            version,
+                                                            force,
+                                                            cc_version,
+                                                            statistics)
+
+                            self.__store_reports(session,
+                                                 report_dir,
+                                                 source_root,
                                                  run_id,
-                                                 durations)
-
-                    store_handler.finishCheckerRun(session, run_id)
-
-                    session.commit()
+                                                 file_path_to_id,
+                                                 run_history_time,
+                                                 self.__context.severity_map,
+                                                 wrong_src_code_comments,
+                                                 skip_handler,
+                                                 checkers)
+
+                            store_handler.setRunDuration(session,
+                                                         run_id,
+                                                         durations)
+
+                            store_handler.finishCheckerRun(session, run_id)
+
+                            session.commit()
+                            break
+                    except Exception as ex:
```

Comment:
What if some other exception was thrown? Plist parse error or something non database related it will still try to do the store three times.

Reply:
We handle plist parsing errors at here:
https://github.com/Ericsson/codechecker/blob/74540cb744606992eefa37363b695573994a8ae4/web/server/codechecker_server/api/report_server.py#L2519-L2522

By the way you are right, it is still possible that non database related exception will be thrown. The problem is that for example when psycopg2 is used the exception will be the following: `psycopg2.extensions.TransactionRollbackError`. When pg8000 is used the exception will be different: `pg8000.core.ProgrammingError`. How we should handle this use case?

Maybe we can use the following:
```py
driver = host_check.get_postgresql_driver_name()
DeadlockException = Exception
if driver == "psycopg2":
    import psycopg2
    DeadlockException = \
        psycopg2.extensions.TransactionRollbackError
elif driver == "pg8000":
    import pg8000
    DeadlockException = pg8000.core.ProgrammingError
```

And we can catch this kind of exception on the following way:
```py
try:
    session.flush()
except DeadlockException:
    ...
```
What do you think @gyorb?

Reply:
I see the problem.
I wonder why sqlalchemy does not wrap the db drivers exceptions. It should wrap them:
https://stackoverflow.com/questions/58740043/how-do-i-catch-a-psycopg2-errors-uniqueviolation-error-in-a-python-flask-app or only these types or exceptions are not wrapped.
Did you check if any of the core exceptions can catch the exceptions from the drivers?
https://docs.sqlalchemy.org/en/13/core/exceptions.html
I think these are the top level exception classes:
exception sqlalchemy.exc.DBAPIError
exception sqlalchemy.exc.SQLALchemyError

Reply:
@gyorb I checked it but unfortunately it does not work:
```py
import psycopg2
import sqlalchemy

try:
    raise psycopg2.extensions.TransactionRollbackError("psycop2 Error")
except sqlalchemy.exc.SQLAlchemyError as ex:
    LOG.error("SQLAlchemyError: %s", ex)
except sqlalchemy.exc.DBAPIError as ex:
    LOG.error("DBAPIError: %s", ex)
except Exception as ex:
    LOG.error("NOT WORKING")
``` 
It will always print the NOT WORKING error message.  Same for the pg8000.

In case of pg8000 the inheritance is the following: `ProgrammingError` -> `DatabaseError` -> `Error` -> `Exception`. For more info [see](https://github.com/mfenniak/pg8000/blob/master/pg8000/core.py#L307).

In case of psycop2 the inheritance is the following `TransactionRollbackError` -> `OperationalError` -> `DatabaseError` -> `Error` -> `Exception`. For more info [see](https://github.com/psycopg/psycopg2/blob/16b35ac77b9c8749198e8ff4d9561a051d1ad947/doc/src/module.rst#exceptions)

Reply:
That's not good. Let's go with the database driver based exception handling, I would not try to store the results again if there was some other Exception during the report processing, the Exception will occur in the next round too and it will fail again using the server resources.
What do you think?

Reply:
I found a solution how we can catch the exception through slqalchemy. We can use the `sqlalchemy.exc.OperationalError`, `sqlalchemy.exc.ProgrammingError`  classes.